src/OrleansRuntime/GrainDirectory/MultiClusterRegistration/GlobalSingleInstanceResponseTracker.cs

```diff
@@ -0,0 +1,93 @@
+﻿using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.GrainDirectory;
+using Orleans.SystemTargetInterfaces;
+
+namespace Orleans.Runtime.GrainDirectory 
+{
+    /// <summary>
+    /// A class that encapsulates response processing logic.
+    /// It is a promise that fires once it has enough responses to make a determination.
+    /// </summary>
+    internal class GlobalSingleInstanceResponseTracker : TaskCompletionSource<GlobalSingleInstanceResponseTracker.Outcome> {
```

Comment:
The important part is that sometimes we don't want to wait for all responses (if a cluster that owns the grain responds with an activation address, we can immediately continue without waiting for other clusters to respond). That's why I am not calling `Task.WhenAll(...)`, but instead signal task completion as soon as I have sufficient information to continue.